ddl/partition.go

```diff
@@ -494,42 +474,41 @@ func checkPartitionKeysConstraint(sctx sessionctx.Context, partExpr string, idxC
 		return err
 	}
 
-	constraints := make(map[string]struct{})
-	for _, uniqCol := range idxColNames {
-		constraints[uniqCol.Column.Name.L] = struct{}{}
-	}
-
 	// Every unique key on the table must use every column in the table's partitioning expression.
 	// See https://dev.mysql.com/doc/refman/5.7/en/partitioning-limitations-partitioning-keys-unique-keys.html
-	if !checkConstraintIncludePartKey(partCols, constraints) {
+	if !checkUniqueKeyIncludePartKey(partCols, idxColNames) {
 		return ErrUniqueKeyNeedAllFieldsInPf.GenWithStackByArgs("UNIQUE INDEX")
 	}
 	return nil
 }
 
-func extractPartitionColumns(sctx sessionctx.Context, partExpr string, tblInfo *model.TableInfo) ([]string, error) {
+func extractPartitionColumns(sctx sessionctx.Context, partExpr string, tblInfo *model.TableInfo) ([]*expression.Column, error) {
 	e, err := expression.ParseSimpleExprWithTableInfo(sctx, partExpr, tblInfo)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	cols := expression.ExtractColumns(e)
-	partCols := make([]string, 0, len(cols))
-	for _, col := range cols {
-		partCols = append(partCols, col.ColName.L)
-	}
-	return partCols, nil
+	return expression.ExtractColumns(e), nil
 }
 
-// checkConstraintIncludePartKey checks that the partitioning key is included in the constraint.
-func checkConstraintIncludePartKey(partkeys []string, constraints map[string]struct{}) bool {
-	for _, pk := range partkeys {
-		if _, ok := constraints[pk]; !ok {
+// checkUniqueKeyIncludePartKey checks that the partitioning key is included in the constraint.
+func checkUniqueKeyIncludePartKey(partCols []*expression.Column, idxCols []*ast.IndexColName) bool {
+	for _, partCol := range partCols {
+		if !findColumnInColumns(partCol, idxCols) {
 			return false
 		}
 	}
 	return true
 }
 
+func findColumnInColumns(c *expression.Column, cols []*ast.IndexColName) bool {
```

Comment:
s/findColumnInColumns/findColumnInIndexCols ?
Could we put it to ddl/column.go ?

Reply:
Done